### ISSUE TITLE
bcachefs fusemount hangs on file write

### ISSUE THREAD

**Issue by cepelinas9000** [koverstreet/bcachefs-tools#171]
I was playing around with fuse mount using bcachefs tool, but noticed that it hangs when there is write attempt. This behavior persist using latest master (https://github.com/koverstreet/bcachefs-tools/commit/81352a29bc619c5882cbdd23dbdaa83ae9e3baee) and arch linux (qemu virtual machine, just installed bcachefs-tools).

as for reproduction the steps are:

```bash
$ dd if=/dev/zero of=test bs=1M count=1025
$ bcachefs format test
$ mkdir /tmp/f2
$ bcachefs fusemount test /tmp/f2 -d

# (other tty)
$ cd /tmp/f2
$ touch a
$ echo a >> a

# (after echo command only kill -9 helps)

```

**Comment by cepelinas9000** [comment]
After further investigation, noticed call to  closure_sync(&cl);  ( https://github.com/koverstreet/bcachefs-tools/blob/81352a29bc619c5882cbdd23dbdaa83ae9e3baee/cmd_fusemount.c#L586 ) never executes closure_sync_fn to finish transaction.

if I comment out closure sync the fuse filesystem works as long as write is in single call. see code below (tried to call bch2_write same style as in read_aligned):
```diff
iff --git a/cmd_fusemount.c b/cmd_fusemount.c
index 46dbb9d..5f1bdf2 100644
--- a/cmd_fusemount.c
+++ b/cmd_fusemount.c
@@ -582,8 +582,10 @@ static int write_aligned(struct bch_fs *c, subvol_inum inum,
                return -ENOSPC;
        }
 
-       closure_call(&op.cl, bch2_write, NULL, &cl);
-       closure_sync(&cl);
+       bch2_write(&op.cl);
+
+       closure_get(&cl);
+       //closure_sync(&cl);
 
        if (!op.error)
                *written_out = op.written << 9;
```

**Comment by cepelinas9000** [comment]
There where two problems:
1. after doing RTFM in closures.h: the parent closure (variable cl) in aligned_write method must have closure_return call. It is done by using op.end_io callback
2. bcachefs executable on startup creates multiple threads to emulate kernel workqueues, timer threads + other components, but when fuse library enter background mode it does through fork call and only calling thread survives. Obvious solution is to disable entering to background as I think is not a big problem (see patch). As other solutions are to make initialization code to be explicitly called or forcibly call again constructors (i mean using __init_array_start and __init_array_end linker variables) 


```diff
diff --git a/cmd_fusemount.c b/cmd_fusemount.c
index 46dbb9d..36aba78 100644
--- a/cmd_fusemount.c
+++ b/cmd_fusemount.c
@@ -391,6 +391,10 @@ static void bcachefs_fuse_read_endio(struct bio *bio)
 	closure_put(bio->bi_private);
 }
 
+static void  bcachefs_fuse_write_endio(struct bch_write_op *op){
+	closure_return(&op->cl);
+}
+
 struct fuse_align_io {
 	off_t		start;
 	size_t		pad_start;
@@ -566,6 +570,8 @@ static int write_aligned(struct bch_fs *c, subvol_inum inum,
 	closure_init_stack(&cl);
 
 	bch2_write_op_init(&op, c, io_opts); /* XXX reads from op?! */
+
+	op.end_io = bcachefs_fuse_write_endio;
 	op.write_point	= writepoint_hashed(0);
 	op.nr_replicas	= io_opts.data_replicas;
 	op.target	= io_opts.foreground_target;
@@ -1174,6 +1180,15 @@ static void usage(char *argv[])
 	printf("\n");
 }
 
+/*
+ * print messages at least having warning level
+ */
+static void bcachefs_fuse_log_func(enum fuse_log_level level, const char *fmt, va_list ap)
+{
+	if (level < FUSE_LOG_NOTICE){
+		vfprintf(stderr, fmt, ap);
+	}
+}
 int cmd_fusemount(int argc, char *argv[])
 {
 	struct fuse_args args = FUSE_ARGS_INIT(argc, argv);
@@ -1255,7 +1270,21 @@ int cmd_fusemount(int argc, char *argv[])
 	/* This print statement is a trigger for tests. */
 	printf("Fuse mount initialized.\n");
 
-	fuse_daemonize(fuse_opts.foreground);
+	// XXX: fuse_daemonize doing fork, but because there are __attribute__((constructor(xxx))) in code 
+	// as temporary work around force foreground mode
+	// see: https://github.com/libfuse/libfuse/wiki/FAQ#how-should-threads-be-started
+	// see: https://github.com/libfuse/libfuse/blob/a466241b45d1dd0bf685513bdeefd6448b63beb6/lib/helper.c#L269
+
+	//fuse_daemonize(fuse_opts.foreground);
+
+	if (fuse_opts.foreground == 0){
+		printf("FUSE is forced to foreground due to usage of gcc constructors\n");
+		fuse_daemonize(1);
+	}
+
+if (fuse_opts.debug == 0){
+	   fuse_set_log_func(bcachefs_fuse_log_func);
+	}
 
 	ret = fuse_session_loop(se);
 

```

after this patch even pytest tests/test_fuse.py passes :)

**Comment by koverstreet** [comment]
So I see you're interested in the fuse code - nice work on debugging the constructor issue.

For bch_write_op, we were getting away from using a closure for returning completions to avoid some atomic ops; adding a .end_io function and then doing a closure_return from that end_io function doesn't make sense - if you're using op->end_io you wouldn't be using op->cl.parent.

I'm not going to have much time for the fuse code myself, but I'll be happy to review and take your patches - feel free to join the IRC channel (oftc #bcache) if you want to chat.

**Comment by cepelinas9000** [comment]
Granted after your explanation it looks that code should not work.

Now I have question about closures, if I interpret correctly when there is **parent** parameter is in **closure_init** the **cosure_sync(parent)** call will block until all children closures are completed, but when I write following minimal code (at a glance the fuse write_aligned function uses similar pattern):
```c
#include "cmd_pg.h"
#include "cmds.h"
#include "linux/closure.h"


void call0(struct closure *cl){
  printf("call0 child, the cl=%p\n",cl);

  //continue_at(cl,NULL,NULL);
  //closure_return(cl);
}

int cmd_pg(int argc, char *argv[]){

  printf("playground command\n");

  struct closure p,c;


  printf("p=%p c=%p\n",&p,&c);
  closure_init_stack(&p);
  
  
  closure_init(&c,&p);
  closure_call(&c,call0,NULL,&p);
  closure_sync(&c);

  //continue_at(&c,NULL,NULL);
  //closure_put(&c);

  printf("clc pass\n");


  closure_sync(&p);
  return 0;
}
```
my playground program hangs (tried same call sequence on kernel module too). My question is what wrong with this code snippet and how to fix it?

**Comment by koverstreet** [comment]
In your closure_call() you didn't specify a workqueue, so that code runs synchronously - should still work, but probably not what you expect.

call0() needs the closure_return().

**Comment by cepelinas9000** [comment]
I would be glad could somebody (I know, that number is very limited)  provide pointers how to approach write_aligned synchronization.


After  digging **_bch2_write_** I noticed that, there are  two general cases on empty filesystem: 
1 - when data if into inode (inline write) the everything work as expected*; 
2 - when data doesn't fit into inode - after executing closure_call(&op.cl,bch2_write,NULL) the value op.cl.remaining is 1 and any sync will rise (closure_put_after_sub: Assertion `!(flags & ((CLOSURE_DESTRUCTOR|CLOSURE_WAITING|CLOSURE_RUNNING) << 1))' failed.)
 \* bch2_write closure, never calls closure_return - I suppose it is expected behavior as it will fall on op.end_io function (wild guess here)

Thus digging deeper I noticed there is following call sequence for second case (note some line numbers can be slightly different) :
1. (thread 1) closure_call(&op.cl,bch2_write,NULL) - cmd_fusemount.c:585 entering closure, cl.remaining=0x40000001 as it is excepted value;
2. (thread 1) closure_get(bio->bi_private) - line libbcachefs/io_write.c:1476 , bio→bi_private is same closure, value op.cl.remaining=0x40000002
3. (thread 1) calls continue_at(&op->cl, bch2_write_index, NULL); - line libbcachefs/io_write.c:1503, now got value **op.cl.remaining=1 !!!**
4. (thread 1) closure_call(&op.cl,bch2_write,NULL) returns and continues to rest method;if closing sequence is called (closure_sync, closure_continoue(&op.cl,NULL,NULL), then if assertion fail due to op.cl.remaining == 1
6. (thread 2 "aio_completion" something one thread yield later [it can be from 10uS to 20uS] on my system from step 4, it is hard to get direct breakpoint due to multiple fuse write calls), - libbcachefs/io_write.c:677 (bch2_write_endio), it fixes value to from 1 to 0x40000001

**extra note!** when stepping with gdb and forgetting scheduler-locking, there other case when op.cl.remaining goes down to 0, and after some time goes to 0x40000001




The synchronous code snippet illustrating sequence using single thread (even it is synchronous, I think :) kept same call sequence):
```c

#include "cmd_pg.h"
#include "cmds.h"
#include "linux/closure.h"

void call1(struct closure *cl){
  //TODO: nothing
  printf("call1 cl=%p remaining=%08x\n",cl,atomic_read(&cl->remaining));
}

void call0(struct closure *cl){

 printf("call0 cl=%p remaining=%08x\n",cl,atomic_read(&cl->remaining));

 closure_get(cl);
 continue_at(cl,call1,NULL);

}


int cmd_pg(int argc, char *argv[]){

  printf("playground command\n");

  struct closure p;

  closure_call(&p,call0,NULL,NULL);
  printf("after closure_call cl.remaining=%x\n",atomic_read(&p.remaining));

  if (argc > 1 && strcmp(argv[1],"fast") == 0){ // call closure_put before sync
      closure_put(&p); 
  }

  printf("after closure_put (if uncommented) cl.remaining=%x\n",atomic_read(&p.remaining));


  closure_sync(&p);

  //closure_put(&p); // in case of cmd_fusemount write_aligned function it is called after closure_sync (thread 1) in (thread 2) "aio_completion" thread

  continue_at(&p,NULL,NULL);
  printf("cl=%x\n",atomic_read(&p.remaining));

  return 0;
}

```

---------------
pointers how I debugged it (changed compilation flags to "-og -ggdb3" ):

```gdb
gdb bcachefs

# the cmd_fusemount.c lines will be different due to me playing around
1)
(gdb) b cmd_fusemount.c:579 # just before closure_call
(gdb) b cmd_fusemount.c:588 # whatever, just after closure_call

(gdb) r fusemount testfs /tmp/f2 -d # start with some formatted filesytem (with almost empty image)
(shell) $ cp ~/something_about_15kb /tmp/f2

2) after hit first breakpoint, lock scheduler threads

(gdb) set scheduler-locking on # disable execution for other threads
(gdb) watch op.cl.remaining.counter # watch closure remaining variable
...continue until hit second breakpoint...
...you can safely disable second breakpoint...

(gdb) set scheduler-locking off # resume execution for all threads
(gdb) b closure_put if cl = 0x7fffffffd060 # address of op.cl
(gdb) c # wait for closure_put hit

```

**Comment by cepelinas9000** [comment]
I think bug can be closed, as after merging PR #191 does not hangs on write.
